File: src/parser_utils/output/source.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

type t = { buffer: Buffer.t }

let create () = { buffer = Buffer.create 127 }

let add_string str src =
  Buffer.add_string src.buffer str;
  src

let add_identifier str src = add_string str src

(* TODO: Remove any right trailing whitespace *)
let add_newline source =
  Buffer.add_string source.buffer "\n";
  source

let add_space num b = add_string (String.make num ' ') b

let contents b = Buffer.contents b.buffer

File: rust_port/crates/flow_parser_utils_output/src/source.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#[derive(Default)]
pub struct Source {
    buffer: String,
}

impl Source {
    pub fn create() -> Self {
        Self {
            buffer: String::with_capacity(127),
        }
    }

    pub fn add_string(&mut self, s: &str) -> &mut Self {
        self.buffer.push_str(s);
        self
    }

    pub fn add_identifier(&mut self, s: &str) -> &mut Self {
        self.add_string(s)
    }

    // TODO: Remove any right trailing whitespace
    pub fn add_newline(&mut self) -> &mut Self {
        self.buffer.push('\n');
        self
    }

    pub fn add_space(&mut self, n: usize) -> &mut Self {
        let spaces = " ".repeat(n);
        self.add_string(&spaces)
    }

    pub fn contents(self) -> String {
        self.buffer
    }
}
